File: .claude/skills/level0/lib/paths.js
// The path a rule scopes on. The client hands the door an absolute path, and
// Vale and the judge both read the path the repo root holds.
// [[spec/design_output/level0#the-path-a-rule-reads]]

import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

export function relativeTo(root, path) {
  const said = slashed(path);
  const at = slashed(root).replace(/\/+$/, "");
  if (!at) return said;

  const under = `${at}/`;
  if (said.toLowerCase().startsWith(under.toLowerCase()))
    return said.slice(under.length);
  return said;
}

export function matches(glob, path) {
  return globOf(glob).test(slashed(path));
}

// A script runs its main where node runs that file, so a test importing it registers its cases and meets no exit. [[spec/design_output/doors#a-script-guards-its-main]]
export function runsHere(url, argv) {
  return Boolean(argv[1]) && resolve(argv[1]) === fileURLToPath(url);
}

// [[spec/design_output/schema#the-underscore-parks-a-draft]]
export function isDraft(path) {
  return slashed(path)
    .split("/")
    .some((part) => part.startsWith("_"));
}

function slashed(said) {
  return String(said ?? "")
    .split("\\")
    .join("/");
}

function globOf(glob) {
  const said = String(glob ?? "")
    .split(/(\*\*\/|\*\*|\*|\?)/)
    .map((part) => {
      if (part === "**/") return "(?:.*/)?";
      if (part === "**") return ".*";
      if (part === "*") return "[^/]*";
      if (part === "?") return "[^/]";
      return part.replace(/[.+^${}()|[\]\\]/g, "\\$&");
    })
    .join("");
  return new RegExp(`^${said}$`);
}

File: src/lsp/paths.go
// The path a rule scopes on. A caller hands an absolute path, and every rule
// reads the path the repo root holds.
// [[spec/design_output/level0#the-path-a-rule-reads]]
package main

import (
	"regexp"
	"strings"
)

func slashed(said string) string {
	return strings.ReplaceAll(said, "\\", "/")
}

func relativeTo(root, path string) string {
	said := slashed(path)
	at := strings.TrimRight(slashed(root), "/")
	if at == "" {
		return said
	}
	under := at + "/"
	if strings.HasPrefix(strings.ToLower(said), strings.ToLower(under)) {
		return said[len(under):]
	}
	return said
}

// [[spec/design_output/schema#the-underscore-parks-a-draft]]
func isDraft(path string) bool {
	for _, part := range strings.Split(slashed(path), "/") {
		if strings.HasPrefix(part, "_") {
			return true
		}
	}
	return false
}

var globParts = regexp.MustCompile(`(\*\*|\*|\?)`)

func matches(glob, path string) bool {
	return globOf(glob).MatchString(slashed(path))
}

func globOf(glob string) *regexp.Regexp {
	said := &strings.Builder{}
	said.WriteString("^")
	for _, part := range splitKeeping(glob) {
		switch part {
		case "**":
			said.WriteString(".*")
		case "*":
			said.WriteString("[^/]*")
		case "?":
			said.WriteString("[^/]")
		default:
			said.WriteString(regexp.QuoteMeta(part))
		}
	}
	said.WriteString("$")
	found, err := regexp.Compile(said.String())
	if err != nil {
		return regexp.MustCompile(`^$`)
	}
	return found
}

// [[spec/design_output/level0#the-path-a-rule-reads]]
func splitKeeping(said string) []string {
	out := []string{}
	at := 0
	for _, where := range globParts.FindAllStringIndex(said, -1) {
		out = append(out, said[at:where[0]], said[where[0]:where[1]])
		at = where[1]
	}
	return append(out, said[at:])
}
